client/src/logic/impl/MapNavigatorImpl.ts

```diff
@@ -1,19 +1,92 @@
 import { MapNode } from "../../types/graph/MapNode";
-import { MapNodeFilter } from "../../types/MapNodeFilter";
-import { NavigationDirections } from "../../types/NavigationDirections";
-import { NavigationNode } from "../../types/NavigationNode";
+import { MapNodeFilter } from "../../types/roomsearch/MapNodeFilter";
+import { NavigationDirections } from "../../types/navigation/NavigationDirections";
+import { NavigationNode } from "../../types/navigation/NavigationNode";
 import { MapNavigator } from "../interfaces/MapNavigator";
+import { Graph } from "../interfaces/Graph";
+import { NeighbourConnection } from "../../types/graph/NeighbourConnection";
+import { Constants } from "../../utils/Constants";
 
 export class MapNavigatorImpl implements MapNavigator {
-    
+    private graph: Graph;
+
+    constructor(graph: Graph) {
+        this.graph = graph;
+    }
+
     findShortestPath(startNodeId: number, endNodeFilter: MapNodeFilter): NavigationDirections {
         let path = this.findPathToNearestNode(startNodeId, endNodeFilter);
         let navNodes = this.convertPathToNavigationNodes(path);
         return this.splitNavigationPathIntoSteps(navNodes);
     }
 
-    private findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
-        throw new Error("Method not implemented.");
+    /*private*/ findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
+
+        if (typeof this.graph.getNode(startNodeId) === "undefined") {
+            throw new Error("Start location not valid.");
+        }
+
+        let path: MapNode[] = [];
+        let neighbours: NeighbourConnection[];
+
+        let parents = new Map<number, number>(); //childId, parentId
+        let bestDistances = new Map<number, number>(); //nodeId, distance
+
+        let SortedSet = require('sorted-set');
+        let unvisitedNodes = new SortedSet({
+            hash(item: any) {return item.nodeId},
+            compare(a: any, b: any) {return a.distance - b.distance;}
+        });
+
+        let currentDistance: number;
+        let currentNodeId: number;
+        let allVisited: boolean;
+
+        bestDistances.set(startNodeId, 0);
+        unvisitedNodes.add({nodeId: startNodeId, distance: 0});
+
+        while (true) {
+            currentDistance = Constants.infinity;
+            allVisited = true;
+
+            if (unvisitedNodes.length > 0) {
+                currentNodeId = unvisitedNodes.head().nodeId;
+                currentDistance = unvisitedNodes.head().distance;
+                allVisited = false;
+            }
```

Comment:
After conforming the while to stop when the set is empty, this will not be needed. Just read the head (the first two lines in if are good, but if condition should be removed because it will always be true.